File: app/src/main/java/com/arturogutierrez/openticator/domain/navigator/drawer/NavigationDrawerPresenter.java
package com.arturogutierrez.openticator.domain.navigator.drawer;

import com.arturogutierrez.openticator.domain.category.CategorySelector;
import com.arturogutierrez.openticator.domain.category.interactor.GetCategoriesInteractor;
import com.arturogutierrez.openticator.domain.category.model.Category;
import com.arturogutierrez.openticator.interactor.DefaultSubscriber;
import com.arturogutierrez.openticator.view.presenter.Presenter;
import java.util.List;
import javax.inject.Inject;

public class NavigationDrawerPresenter extends DefaultSubscriber<List<Category>>
    implements Presenter {

  private final GetCategoriesInteractor getCategoriesInteractor;
  private final CategorySelector categorySelector;

  private NavigationDrawerView view;
  private List<Category> categories;

  @Inject
  public NavigationDrawerPresenter(GetCategoriesInteractor getCategoriesInteractor,
      CategorySelector categorySelector) {
    this.getCategoriesInteractor = getCategoriesInteractor;
    this.categorySelector = categorySelector;
  }

  public void setView(NavigationDrawerView view) {
    this.view = view;
  }

  @Override
  public void resume() {
    getCategoriesInteractor.execute(this);
  }

  @Override
  public void pause() {
    getCategoriesInteractor.unsubscribe();
  }

  @Override
  public void destroy() {
    getCategoriesInteractor.unsubscribe();
  }

  @Override
  public void onNext(List<Category> categories) {
    this.categories = categories;
    view.renderCategories(categories);
  }

  public void categoryItemClicked(int position) {
    if (position < categories.size()) {
      categorySelector.setSelectedCategory(categories.get(position));
    }

    view.dismissDrawer();
  }

  public void allCategoriesItemClicked() {
    categorySelector.removeSelectedCategory();
    view.dismissDrawer();
  }
}

File: app/src/main/kotlin/com/arturogutierrez/openticator/domain/navigator/drawer/NavigationDrawerPresenter.kt
package com.arturogutierrez.openticator.domain.navigator.drawer

import com.arturogutierrez.openticator.domain.category.CategorySelector
import com.arturogutierrez.openticator.domain.category.interactor.GetCategoriesInteractor
import com.arturogutierrez.openticator.domain.category.model.Category
import com.arturogutierrez.openticator.interactor.DefaultFlowableObserver
import com.arturogutierrez.openticator.view.presenter.Presenter
import javax.inject.Inject

class NavigationDrawerPresenter @Inject constructor(
    private val getCategoriesInteractor: GetCategoriesInteractor,
    private val categorySelector: CategorySelector)
  : Presenter<NavigationDrawerView>() {

  // TODO: Presenter shouldn't has state
  private var categories: List<Category> = emptyList()

  override fun resume() {
    val params = GetCategoriesInteractor.EmptyParams
    getCategoriesInteractor.execute(params, object : DefaultFlowableObserver<List<Category>>() {
      override fun onNext(t: List<Category>) {
        onFetchCategories(t)
      }
    })
  }

  override fun pause() {
    getCategoriesInteractor.dispose()
  }

  override fun destroy() {
    getCategoriesInteractor.dispose()
  }

  private fun onFetchCategories(categories: List<Category>) {
    this.categories = categories
    view?.renderCategories(categories)
  }

  fun categoryItemClicked(position: Int) {
    if (position < categories.size) {
      categorySelector.setSelectedCategory(categories[position])
    }

    view?.dismissDrawer()
  }

  fun allCategoriesItemClicked() {
    categorySelector.removeSelectedCategory()
    view?.dismissDrawer()
  }
}
